lib/internal/fs/promises.js

```diff
@@ -316,25 +315,46 @@ async function readFileHandle(filehandle, options) {
   if (size > kIoMaxLength)
     throw new ERR_FS_FILE_TOO_LARGE(size);
 
-  const chunks = [];
-  let isFirstChunk = true;
-  const firstChunkSize = size === 0 ? kReadFileMaxChunkSize : size;
-  const chunkSize = MathMin(firstChunkSize, kReadFileMaxChunkSize);
   let endOfFile = false;
+  let totalRead = 0;
+  const noSize = size === 0;
+  const buffers = [];
+  const fullBuffer = noSize ? undefined : Buffer.allocUnsafeSlow(size);
   do {
     if (signal?.aborted) {
       throw lazyDOMException('The operation was aborted', 'AbortError');
     }
-    const buf = Buffer.alloc(isFirstChunk ? firstChunkSize : chunkSize);
-    const { bytesRead, buffer } =
-      await read(filehandle, buf, 0, buf.length, -1);
-    endOfFile = bytesRead === 0;
```

Comment:
nit: it looks like it should be enough to change this line to:
```js
    totalRead += bytesRead;
    endOfFile = bytesRead === 0 || totalRead === size;
```
in order to get the same performance improvement as in this PR (at least that's what I saw on my machine). Another useful optimization here is `Buffer.allocUnsafeSlow`. So, maybe it's worth considering keeping only these 4 lines of changes, but with the same result.